File: challenges/07_type_hinting/python/type_helper.py
"""
Type Hinting Example

This file demonstrates how to use GitHub Copilot for type hinting.
Each section contains types and protocols that you can ask questions about.
"""

from typing import TypeVar, Generic, Dict, List, Optional, Union, Protocol, TypedDict, NewType, Any
from dataclasses import dataclass
from datetime import datetime
import uuid

# Basic Types and Classes
UserId = NewType('UserId', str)
Role = Union['admin', 'user', 'guest']

@dataclass
class UserPreferences:
    theme: str
    notifications: bool
    language: str

@dataclass
class User:
    id: UserId
    name: str
    email: str
    role: Role
    preferences: Optional[UserPreferences]
    created_at: datetime
    updated_at: datetime

# Generic Types and Protocols
T = TypeVar('T', bound='HasId')

class HasId(Protocol):
    id: str

class Repository(Generic[T]):
    """
    TODO: Ask Copilot questions about Repository implementation:
    - How to handle type constraints?
    - What about error handling?
    - Should we use generics?
    - How to implement caching?
    - What about transaction support?
    """
    def __init__(self, name: str, validate: callable):
        self.name = name
        self.validate = validate
        self.store: Dict[str, T] = {}

    async def find_by_id(self, id: str) -> Optional[T]:
        return self.store.get(id)

    async def find_all(self) -> List[T]:
        return list(self.store.values())

    async def create(self, item: Dict) -> T:
        id = str(uuid.uuid4())
        new_item = {**item, 'id': id}
        if not self.validate(new_item):
            raise ValueError('Invalid item')
        self.store[id] = new_item  # type: ignore
        return new_item  # type: ignore

    async def update(self, id: str, item: Dict) -> T:
        existing = self.store.get(id)
        if not existing:
            raise ValueError('Item not found')
        updated = {**existing, **item}  # type: ignore
        if not self.validate(updated):
            raise ValueError('Invalid item')
        self.store[id] = updated  # type: ignore
        return updated  # type: ignore

    async def delete(self, id: str) -> bool:
        if id in self.store:
            del self.store[id]
            return True
        return False

# Union Types and Type Guards
class Success(TypedDict):
    success: bool
    data: Any

class Failure(TypedDict):
    success: bool
    error: Exception

Result = Union[Success, Failure]

class TypeHelper:
    """Collection of type-safe utilities"""

    @staticmethod
    async def handle_result(promise: callable) -> Result:
        """
        TODO: Ask Copilot questions about Result handling:
        - How to handle async operations?
        - What about type inference?
        - Should we use type guards?
        - How to chain results?
        - What about error types?
        """
        try:
            data = await promise()
            return {'success': True, 'data': data}
        except Exception as error:
            return {'success': False, 'error': error}

    @staticmethod
    def create_type_guard(check: callable) -> callable:
        """
        TODO: Ask Copilot questions about type guards:
        - When to use type predicates?
        - How to narrow union types?
        - What about branded types?
        - Should we use assertions?
        - How to handle unknown types?
        """
        def guard(value: Any) -> bool:
            try:
                return check(value)
            except:
                return False
        return guard

    @staticmethod
    def create_type_mapper(transform: callable) -> callable:
        """
        TODO: Ask Copilot questions about type mapping:
        - How to transform types?
        - What about conditional types?
        - Should we use template literals?
        - How to handle recursion?
        - What about type inference?
        """
        def mapper(items: List[Any]) -> List[Any]:
            return [transform(item) for item in items]
        return mapper

async def run_examples():
    """Example usage of the type helpers"""
    # Create a user repository
    def validate_user(user: Dict) -> bool:
        return (
            isinstance(user.get('id'), str) and
            isinstance(user.get('name'), str) and
            isinstance(user.get('email'), str) and
            user.get('role') in ['admin', 'user', 'guest']
        )

    user_repo = Repository[User]('users', validate_user)

    # Test result handling
    result = await TypeHelper.handle_result(
        lambda: user_repo.create({
            'name': 'Alice',
            'email': 'alice@example.com',
            'role': 'user',
            'created_at': datetime.now(),
            'updated_at': datetime.now()
        })
    )

    if result['success']:
        print('Created user:', result['data'])
    else:
        print('Error:', result['error'])

    # Test type guard
    is_user = TypeHelper.create_type_guard(
        lambda value: (
            isinstance(value, dict) and
            'id' in value and
            'name' in value and
            'email' in value
        )
    )

    unknown_value = {'id': '1', 'name': 'Bob', 'email': 'bob@example.com'}
    if is_user(unknown_value):
        print('Valid user:', unknown_value['name'])

    # Test type mapper
    to_user_dto = TypeHelper.create_type_mapper(
        lambda user: {'name': user['name'], 'email': user['email']}
    )

    users = await user_repo.find_all()
    dtos = to_user_dto(users)
    print('User DTOs:', dtos)

if __name__ == '__main__':
    import asyncio
    asyncio.run(run_examples())

# Practice exercises:
# TODO: Ask Copilot questions about these patterns
# 1. Protocol types
# 2. Type variables
# 3. Overload methods
# 4. Generic types
# 5. Type aliases 

File: challenges/07_type_hinting/javascript/type_helper.js
/**
 * Type Hinting Example
 * 
 * This file demonstrates how to use GitHub Copilot for type hinting in JavaScript using JSDoc.
 * Each section contains types and protocols that you can ask questions about.
 */

// Basic Types using JSDoc
/**
 * @typedef {Object} UserPreferences
 * @property {string} theme - The user's preferred theme
 * @property {boolean} notifications - Whether notifications are enabled
 * @property {string} language - The user's preferred language
 */

/**
 * @typedef {Object} User
 * @property {string} id - Unique identifier
 * @property {string} name - User's full name
 * @property {string} email - User's email address
 * @property {'admin' | 'user' | 'guest'} role - User's role
 * @property {UserPreferences} [preferences] - Optional user preferences
 * @property {Date} createdAt - Account creation date
 * @property {Date} updatedAt - Last update date
 */

// Generic Types and Constraints using JSDoc
/**
 * @template T
 * Repository class for managing entities with IDs
 */
class Repository {
    /**
     * TODO: Ask Copilot questions about Repository implementation:
     * - How to handle type constraints?
     * - What about error handling?
     * - Should we use generics?
     * - How to implement caching?
     * - What about transaction support?
     */
    constructor(validate) {
        this.store = new Map();
        this.validate = validate;
    }

    /**
     * Find an entity by ID
     * @param {string} id - Entity ID
     * @returns {Promise<T | undefined>} Found entity or undefined
     */
    async findById(id) {
        return this.store.get(id);
    }

    /**
     * Find all entities
     * @returns {Promise<T[]>} Array of all entities
     */
    async findAll() {
        return Array.from(this.store.values());
    }

    /**
     * Create a new entity
     * @param {Omit<T, 'id'>} item - Entity data without ID
     * @returns {Promise<T>} Created entity
     * @throws {Error} If validation fails
     */
    async create(item) {
        const id = crypto.randomUUID();
        const newItem = { ...item, id };
        if (!this.validate(newItem)) {
            throw new Error('Invalid item');
        }
        this.store.set(id, newItem);
        return newItem;
    }

    /**
     * Update an existing entity
     * @param {string} id - Entity ID
     * @param {Partial<T>} item - Partial entity data
     * @returns {Promise<T>} Updated entity
     * @throws {Error} If entity not found or validation fails
     */
    async update(id, item) {
        const existing = this.store.get(id);
        if (!existing) {
            throw new Error('Item not found');
        }
        const updated = { ...existing, ...item };
        if (!this.validate(updated)) {
            throw new Error('Invalid item');
        }
        this.store.set(id, updated);
        return updated;
    }

    /**
     * Delete an entity
     * @param {string} id - Entity ID
     * @returns {Promise<boolean>} Whether the entity was deleted
     */
    async delete(id) {
        return this.store.delete(id);
    }
}

// Union Types and Type Guards using JSDoc
/**
 * @template T
 * @typedef {Object} Success
 * @property {true} success - Operation succeeded
 * @property {T} data - Result data
 */

/**
 * @typedef {Object} Failure
 * @property {false} success - Operation failed
 * @property {Error} error - Error information
 */

/**
 * @template T
 * @typedef {Success<T> | Failure} Result
 */

class TypeHelper {
    /**
     * TODO: Ask Copilot questions about Result handling:
     * - How to handle async operations?
     * - What about type inference?
     * - Should we use type guards?
     * - How to chain results?
     * - What about error types?
     * 
     * @template T
     * @param {() => Promise<T>} promise - Promise to handle
     * @returns {Promise<Result<T>>} Result object
     */
    static async handleResult(promise) {
        try {
            const data = await promise();
            return { success: true, data };
        } catch (error) {
            return { success: false, error };
        }
    }

    /**
     * TODO: Ask Copilot questions about type guards:
     * - When to use type predicates?
     * - How to narrow union types?
     * - What about branded types?
     * - Should we use assertions?
     * - How to handle unknown types?
     * 
     * @template T
     * @param {(value: unknown) => boolean} check - Type check function
     * @returns {(value: unknown) => value is T} Type guard function
     */
    static createTypeGuard(check) {
        return (value) => {
            try {
                return check(value);
            } catch {
                return false;
            }
        };
    }

    /**
     * TODO: Ask Copilot questions about type mapping:
     * - How to transform types?
     * - What about conditional types?
     * - Should we use template literals?
     * - How to handle recursion?
     * - What about type inference?
     * 
     * @template T,U
     * @param {(value: T) => U} transform - Transform function
     * @returns {(items: T[]) => U[]} Array transform function
     */
    static createTypeMapper(transform) {
        return (items) => items.map(transform);
    }
}

// Example usage
async function runExamples() {
    // Create a user repository
    const validateUser = (user) => {
        return (
            typeof user.id === 'string' &&
            typeof user.name === 'string' &&
            typeof user.email === 'string' &&
            ['admin', 'user', 'guest'].includes(user.role)
        );
    };

    const userRepo = new Repository(validateUser);

    // Test result handling
    const result = await TypeHelper.handleResult(() =>
        userRepo.create({
            name: 'Alice',
            email: 'alice@example.com',
            role: 'user',
            createdAt: new Date(),
            updatedAt: new Date()
        })
    );

    if (result.success) {
        console.log('Created user:', result.data);
    } else {
        console.log('Error:', result.error);
    }

    // Test type guard
    const isUser = TypeHelper.createTypeGuard((value) => {
        return (
            typeof value === 'object' &&
            value !== null &&
            'id' in value &&
            'name' in value &&
            'email' in value
        );
    });

    const unknownValue = { id: '1', name: 'Bob', email: 'bob@example.com' };
    if (isUser(unknownValue)) {
        console.log('Valid user:', unknownValue.name);
    }

    // Test type mapper
    const toUserDto = TypeHelper.createTypeMapper(
        user => ({ name: user.name, email: user.email })
    );

    const users = await userRepo.findAll();
    const dtos = toUserDto(users);
    console.log('User DTOs:', dtos);
}

// Practice exercises:
// TODO: Ask Copilot questions about these patterns
// 1. JSDoc type definitions
// 2. Type checking functions
// 3. Method overloading with JSDoc
// 4. Generic functions
// 5. Type assertions

module.exports = { Repository, TypeHelper }; 